src/decoder/lattice-incremental-decoder.cc

```diff
@@ -0,0 +1,1199 @@
+
+  KALDI_VLOG(4) << "init:" << num_toks_ / 2 + 3
+                << " buckets:" << tok_map.bucket_count()
+                << " load:" << tok_map.load_factor()
+                << " max:" << tok_map.max_load_factor();
+  // Create initial_arc for later appending with the previous chunk
+  if (create_initial_state) {
+    for (Token *tok = active_toks_[frame_begin].toks; tok != NULL; tok = tok->next) {
+      StateId cur_state = tok_map[tok];
+      // state_label_map_ is construct during create_final_state
+      int32 id = state_label_map_.find(tok)->second; // it should exist
+      // TODO: calculate alpha but not use tot_cost
+      BaseFloat cost_offset = tok->tot_cost;
+      state_label_forward_prob_[id] = tok->tot_cost;
+      Arc arc(0, id, Weight(0, cost_offset), cur_state);
+      ofst->AddArc(begin_state, arc);
+    }
+  }
+  // Now create all arcs.
+  for (int32 f = frame_begin; f <= frame_end; f++) {
+    for (Token *tok = active_toks_[f].toks; tok != NULL; tok = tok->next) {
+      StateId cur_state = tok_map[tok];
+      for (ForwardLinkT *l = tok->links; l != NULL; l = l->next) {
+        if (f == frame_begin && create_initial_state && l->ilabel == 0)
+          continue; // has existed in the last chunk
+        if (f == frame_end && create_final_state && l->ilabel != 0)
+          continue; // will exist in the next chunk
+        typename unordered_map<Token *, StateId>::const_iterator iter =
+            tok_map.find(l->next_tok);
+        StateId nextstate = iter->second;
+        KALDI_ASSERT(iter != tok_map.end());
+        BaseFloat cost_offset = 0.0;
+        if (l->ilabel != 0) { // emitting..
+          KALDI_ASSERT(f >= 0 && f < cost_offsets_.size());
+          cost_offset = cost_offsets_[f];
+        }
+        Arc arc(l->ilabel, l->olabel,
+                Weight(l->graph_cost, l->acoustic_cost - cost_offset), nextstate);
+        ofst->AddArc(cur_state, arc);
+      }
+      if (f == frame_end) {
+        if (use_final_probs && !final_costs.empty()) {
+          typename unordered_map<Token *, BaseFloat>::const_iterator iter =
+              final_costs.find(tok);
+          if (iter != final_costs.end())
+            ofst->SetFinal(cur_state, LatticeWeight(iter->second, 0));
+        } else {
+          ofst->SetFinal(cur_state, LatticeWeight::One());
+        }
+      }
+    }
+  }
+  // Create final_arc for later appending with the next chunk
+  if (create_final_state) {
+    StateId end_state = ofst->AddState(); // final-state for the chunk
+    ofst->SetFinal(end_state, Weight::One());
+
+    state_label_map_.clear();
+    state_label_map_.reserve(std::min((int32)1e5, config_.max_active));
+    for (Token *tok = active_toks_[frame_end].toks; tok != NULL; tok = tok->next) {
+      StateId cur_state = tok_map[tok];
+      int32 id = state_label_avilable_idx_++;
+      state_label_map_[tok] = id;
+      Weight final_weight =
+          (!decoding_finalized_ && ofst->Final(cur_state) == Weight::Zero())
```

Comment:
This doesn't look like the right logic for deciding the final weights (plus it's an extremely unclear expression; I suspect you meant || instead of &&, but  that wouldn't fix it.
The logic should be more similar to around line 1145.  If we are using final-probs (e.g. it's the last frame of decoding and the user specified use_final_probs) *and * at least one active state was final, then you should use the final probs for all states, even if most were zero.

Reply:
... and use an if-statement rather than a boolean expression like this, it will  be clearer (I think you even confused yourself here).

Reply:
Thanks. I will re-work on this part. 
Actually, these boolean expression is of no use, since the value of ofst->Final(cur_state) has been taken care in line 1145.  we can just make 
Weight final_weight = ofst->Final(cur_state);

Ya, I should merge these two parts and set final weight and states in one place.

Reply:
Thanks. I will re-work on this part. 
Actually, these boolean expression is of no use, since the value of ofst->Final(cur_state) has been taken care in line 1145.  we can just do:
Weight final_weight = ofst->Final(cur_state);

Ya, I should merge these two parts and set the final weight and states in one place.